src/mscorlib/src/System/IO/Stream.cs

```diff
@@ -111,36 +111,7 @@ public virtual int WriteTimeout
 
         public Task CopyToAsync(Stream destination)
         {
-            int bufferSize = _DefaultCopyBufferSize;
-
-            if (CanSeek)
-            {
-                long length = Length;
-                long position = Position;
-                if (length <= position) // Handles negative overflows
-                {
-                    // If we go down this branch, it means there are
-                    // no bytes left in this stream.
-
-                    // Ideally we would just return Task.CompletedTask here,
-                    // but CopyToAsync(Stream, int, CancellationToken) was already
-                    // virtual at the time this optimization was introduced. So
-                    // if it does things like argument validation (checking if destination
-                    // is null and throwing an exception), then await fooStream.CopyToAsync(null)
-                    // would no longer throw if there were no bytes left. On the other hand,
-                    // we also can't roll our own argument validation and return Task.CompletedTask,
-                    // because it would be a breaking change if the stream's override didn't throw before,
-                    // or in a different order. So for simplicity, we just set the bufferSize to 1
-                    // (not 0 since the default implementation throws for 0) and forward to the virtual method.
-                    bufferSize = 1;
-                }
-                else
-                {
-                    long remaining = length - position;
-                    if (remaining > 0) // In the case of a positive overflow, stick to the default size
-                        bufferSize = (int)Math.Min(bufferSize, remaining);
-                }
-            }
+            var bufferSize = GetBufferSize(Length, Position, CanSeek);
```

Comment:
I don't think this works.  In the original implementation, Length and Position are guarded behind the call to CanSeek; that's no longer the case.  And on most streams where CanSeek == false, Length and Position will throw.

Reply:
Ah! I'll move them back...